### DISCUSSION TITLE
Keycloak.X: New store

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#9861] [Archive]
Feedback and general discussions around the new store that is being built for Keycloak.X.

Other relevant discussions:

- https://github.com/keycloak/keycloak/discussions/9596
- https://github.com/keycloak/keycloak/discussions/9619

**Comment by xgp** [comment]
Can you please post the link to the example implementation based on ConcurrentHashMap?

**Reply by mhajas** [reply]
Sure, here is what we have so far:

- [ConcurrentHashMap](https://github.com/keycloak/keycloak/tree/main/model/map/src/main/java/org/keycloak/models/map/storage/chm)
- [JPA](https://github.com/keycloak/keycloak/tree/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa)
- [HotRod](https://github.com/keycloak/keycloak/tree/main/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod)

**Comment by andreaTP** [comment]
I have just watched an awesome presentation regarding the new Store at DevCon.cz and I would be happy to understand a few additional details (please feel free to link me to other documents in case I'm missing those).

In the talk have been mentioned that version `N` and `N+1` of the Store needs to be compatible for performing a zero downtime upgrade. This open a few questions:

- How do you guarantee that upgrades will be always "sequential"? How would the new storage deal with "non-sequential" upgrades?
- Reading [this](https://github.com/keycloak/keycloak-community/pull/321) I assume that the schema upgrades will be checked to be compatible using Protobuf, one of the common pitfalls is the exponential growth  and increase of complexity in the protocol to preserve the compatibility. Do you expect the schema to stay pretty much stable or do you have any guard from this risk?
- Not being able to perform breaking changes(or introducing the need of an intermediate release to do so) can slow down quite significantly the development cycle, have we considered possible mitigation strategies to counterbalance this effect, or we simply consider it acceptable?

Thanks a lot!

**Reply by hmlnarik** [reply]
Thank you for the questions!

> In the talk have been mentioned that version `N` and `N+1` of the Store needs to be compatible for performing a zero downtime upgrade. This open a few questions:
> 
> * How do you guarantee that upgrades will be always "sequential"? How would the new storage deal with "non-sequential" upgrades?

The proposal counts on arbitrary ordered sequences, please check [here](https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/storage-persistence.md#_otherVersioning).

> * Reading [this](https://github.com/keycloak/keycloak-community/pull/321) I assume that the schema upgrades will be checked to be compatible using Protobuf, one of the common pitfalls is the exponential growth  and increase of complexity in the protocol to preserve the compatibility. Do you expect the schema to stay pretty much stable or do you have any guard from this risk?

We expect the schema to be pretty much stable, leveraging generic attributes as much as possible. Please note that the reference you have used refers to Hot Rod-specific implementation, and that protobuf does not necessarily apply to other storage technologies.

> * Not being able to perform breaking changes(or introducing the need of an intermediate release to do so) can slow down quite significantly the development cycle, have we considered possible mitigation strategies to counterbalance this effect, or we simply consider it acceptable?

The strategy is based on keeping the schema as stable as possible.
